### PR TITLE
Allow custom labels

### PR DESCRIPTION
**Pull request by dj707chen**
Allow adding custom labels.

**Background**
In a situation where a service interacts with multiple providers, a few of them do not have a lot of traffic, so when one of these light providers is down, the alert of the service is not triggered due to the small percentage counted in the failing traffic; So we want to create alerts specific to providers, and I’m searching for ways to identify the metrics entries related to a provider.
Two options currently available:
- Use provider specific prefix when create `MetricsOps` instance, but the metrics of this client will be excluded from a generic Http4S metrics dashboard;
- Add provider specific prefix to classifier when create `org.http4s.client.middleware.Metrics`.

The second option is better than the first option, but not ideal.

**What it does**
This PR enables clients to pass custom labels and corresponding values when create `MetricsOps` and allow alerts targeting those label.

**Code refactor**
There were a few configuration items (prefix, registry, sampleExemplar, responseDurationSecondsHistogramBuckets), to reduce the number of methods, converted the code to builder pattern, then added `withCustomLabelsAndValues` method.

**Example**
Here is an example. If we create a `MetricsOps` as below and pass it to a client `Metrics` middleware
```
  val metricsOpsPaypal = Prometheus.default(registry)
        .withCustomLabelsAndValues(List(("provider", "Paypal"))
        .build
  val paypalClient = BlazeClientBuilder[F]
      .resource
      .map { client => org.http4s.client.middleware.Metrics[F](metricsOpsPaypal)(client) }
```
Now we can create a `Prometheus` rule as below, note, label `status` is set by `Metrics` middleware, `app` by filter, and `provider` is a custom label.
```
- name: ./team1.rules
   rules:
   - alert: PaypalProviderFailed
      expr: |
         100 * (
             sum by(classifier)(rate(http4s:org_http4s_server_request_count_total{app="banking", provider="Paypal", status="5xx",}[2m]))
             /
             sum by(classifier)(rate(http4s:org_http4s_server_request_count_total{app="banking", provider="Paypal"}[2m]))
         ) > 5
      for: 5m
      labels:
         severity: warning
         team: "team-1"
   - alert: ZelleProviderFailed
      ......
```

### CODE REVIEW

File: prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/Prometheus.scala [hunk 1]
```diff
@@ -110,33 +139,72 @@ object Prometheus {
...
+    * @param prefix a prefix that will be added to all metrics
+    * @param responseDurationSecondsHistogramBuckets the buckets of response duration time in second for Histogram.
+    * @param customLabelsAndValues a list of custom labels and corresponding values, will be added to all metrics entries
+    */
+  def metricsOpsWithExemplarsWithCustomLabels[F[_]: Sync](
+      registry: CollectorRegistry,
+      sampleExemplar: F[Option[Map[String, String]]],
+      prefix: String = "org_http4s_server",
+      responseDurationSecondsHistogramBuckets: NonEmptyList[Double] = DefaultHistogramBuckets,
+      customLabelsAndValues: List[(String, String)] = List.empty,
```
Comment (rossabaker) [inline prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/Prometheus.scala:176]: I wonder if we should be considering the Builder Pattern, to reduce the proliferation of 2^n methods for n configurations.

File: prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/Prometheus.scala [hunk 2]
```diff
@@ -193,9 +279,11 @@ object Prometheus {
...
             metrics.abnormalTerminations
               .labels(
-                label(classifier),
-                AbnormalTermination.report(AbnormalTermination.Canceled),
-                label(Option.empty),
+                List(
+                  label(classifier),
+                  AbnormalTermination.report(AbnormalTermination.Canceled),
+                  label(Option.empty),
+                ) ++ customLabels: _*
```
Comment (rossabaker) [inline prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/Prometheus.scala:286]: These lists are small enough it probably doesn't matter, but we could optimize these a bit with a list buffer.

File: prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/Prometheus.scala (hunk 1 above)
Comment (dj707chen) [reply prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/Prometheus.scala]: Good idea, thanks. Refactored to follow the builder pattern.

File: prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/Prometheus.scala (hunk 2 above)
Comment (dj707chen) [reply prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/Prometheus.scala]: Good catch, changed to use the more efficient `prepended` function.

**Review by rossabaker — APPROVED** [review]
Looks good, thanks!

**Pull request merged**